File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.js
//Third fastest
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */


var isBalanced = function(root) {
    let result = true;
    const innerFunc = (node) => {
        if (!node) return 0;
        
        let l = innerFunc(node.left);
        let r = innerFunc(node.right);
        if (Math.abs(l - r) > 1) result = false;
        return Math.max(l, r) + 1
    }
    innerFunc(root)
    return result
};


//Second fastest
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    let isBalanced = true;
    
    function findHeight(root) {
        if (!root) {
            return 0
        }
        if (isBalanced === false) return; // optimisation: early exit

        let leftHeight = findHeight(root.left);
        let rightHeight = findHeight(root.right);
        let diff = Math.abs(leftHeight - rightHeight);

        if (diff > 1) {
            isBalanced = false;
        }
        
        return Math.max(leftHeight, rightHeight) + 1;
    }
    
    findHeight(root);
    return isBalanced;
};

//Fastest
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    let balanced = true;
    
    const depthTrav = (root) => {
        if (!root) return 0;
        if(!balanced) return;
        
        const leftHeight = depthTrav(root.left);
        const rightHeight = depthTrav(root.right);
        const heightDiff = leftHeight - rightHeight;
      
        if (heightDiff > 1 || heightDiff < -1) {
          balanced = false;
        }
        
        return Math.max(leftHeight, rightHeight) + 1;
    }
    
    depthTrav(root);
    return balanced;
}

File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.go
//Third fastest
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	
	return abs(depth(root.Left)-depth(root.Right)) <= 1 && isBalanced(root.Left) && isBalanced(root.Right)
}

func depth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	
	return 1 + max(depth(root.Left), depth(root.Right))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}


//Second fasstest
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	_, ret := helper110(root)
	return ret
}

func helper110(root *TreeNode) (int, bool) {
	if root == nil {
		return 0, true
	}
	left, ret := helper110(root.Left)
	if !ret {
		return 0, false
	}
	right, ret := helper110(root.Right)
	if !ret {
		return 0, false
	}
	if left-right < -1 || left-right > 1 {
		return 0, false
	}
	if left > right {
		return left + 1, true
	}
	return right + 1, true
}

//Fastest
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(node *TreeNode) bool {
	if node == nil {
		return true
	}

    balanced := math.Abs(float64(getHeight(node.Left)-getHeight(node.Right))) <= 1
	return balanced && isBalanced(node.Left) && isBalanced(node.Right)
}

func getHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}

	leftHeight := 1 + getHeight(node.Left)
	rightHeight := 1 + getHeight(node.Right)

	return int(math.Max(float64(leftHeight), float64(rightHeight)))
}